### PR TITLE
Introduce writeMessage and onMessage APIs with data binding

### PR DESCRIPTION
**Pull request by Bhashinee**
## Purpose
Proposal: https://github.com/ballerina-platform/ballerina-standard-library/issues/2761

## Checklist
- [x] Linked to an issue
- [x] Updated the specification
- [ ] Updated the changelog
- [x] Added tests

### CODE REVIEW

**Comment by shafreenAnfar** [comment]
Just to clarify we only data binding and `anydata` support for `readMessage`, `writeMessage` and `onMessage`. Rest of the APIs can remain the same.

File: docs/spec/spec.md [hunk 1]
```diff
@@ -3,53 +3,48 @@
 _Owners_: @shafreenAnfar @bhashinee  
 _Reviewers_: @shafreenAnfar  
 _Created_: 2021/12/09  
-_Updated_: 2022/02/18  
-_Edition_: Swan Lake  
+_Updated_: 2021/12/09
```
Comment (dilanSachi) [inline docs/spec/spec.md:6]:
```suggestion
_Updated_: 2022/05/05  
```

File: native/src/main/java/io/ballerina/stdlib/websocket/WebSocketResourceDispatcher.java [hunk 2]
```diff
@@ -500,38 +503,35 @@ public static void dispatchOnText(WebSocketConnectionInfo connectionInfo, WebSoc
...
-                                }
-                                if (readOnly) {
-                                    record = CloneReadOnly.cloneReadOnly(record);
+                                break;
+                            case TypeTags.UNION_TAG:
+                                if (WebSocketUtil.hasStringType(param)) {
+                                    bValue = CloneWithType.convert(param,
+                                            StringUtils.fromString(stringAggregator.getAggregateString()));
+                                } else {
+                                    bValue = FromJsonStringWithType.fromJsonStringWithType(StringUtils.fromString(
```
Comment (dilanSachi) [inline native/src/main/java/io/ballerina/stdlib/websocket/WebSocketResourceDispatcher.java:517]: we can send the else block to default case right?

File: native/src/main/java/io/ballerina/stdlib/websocket/WebSocketUtil.java [hunk 3]
```diff
@@ -218,6 +222,21 @@ static String getErrorMessage(Throwable err) {
...
+                new String(byteArray, StandardCharsets.UTF_8));
+    }
+
+    public static boolean hasStringType(Type targetType) {
+        List<Type> memberTypes = ((UnionType) targetType).getMemberTypes();
+        return memberTypes.stream().anyMatch(member -> member.getTag() == TypeTags.STRING_TAG);
+    }
+
+    public static boolean hasByteArrayType(Type targetType) {
+        List<Type> memberTypes = ((UnionType) targetType).getMemberTypes();
```
Comment (dilanSachi) [inline native/src/main/java/io/ballerina/stdlib/websocket/WebSocketUtil.java:236]: wont this accept types like `string[]` too?

File: native/src/main/java/io/ballerina/stdlib/websocket/WebSocketResourceDispatcher.java (hunk 2 above)
Comment (Bhashinee) [reply native/src/main/java/io/ballerina/stdlib/websocket/WebSocketResourceDispatcher.java]: Changed the code to fall through.

File: native/src/main/java/io/ballerina/stdlib/websocket/WebSocketUtil.java (hunk 3 above)
Comment (Bhashinee) [reply native/src/main/java/io/ballerina/stdlib/websocket/WebSocketUtil.java:236]: Fixed with 12d7631

**Comment by Bhashinee** [comment]
> Just to clarify we only need data binding and `anydata` support for `readMessage`, `writeMessage` and `onMessage`. Rest of the APIs can remain the same.

Addressed the suggestions. Removed the data binding support from `onTextMessage`, `readTextMessage` and `writeTextMessage` with [8f53776](https://github.com/ballerina-platform/module-ballerina-websocket/pull/587/commits/8f53776c5a0beda8c380e7df3350dcde7aeafc48)

**Review by dilanSachi — APPROVED** [review]
LGTM

**Pull request merged**